app/presenters/citation_display.rb

```diff
@@ -0,0 +1,35 @@
+# Renders an html_safe citation in HTML from a Work.
+#
+# Uses ruby CSL.
+#
+# Assumes the chicago-note-bibliography CSL style and en-US locale -- lazy loads them
+# both globally because they are slow to load.
+#
+# @example
+#
+# CitationDisplay.new(work).display
+
+class CitationDisplay < ViewModel
+  valid_model_type_names 'Work'
+
+  def display
+    attributes = CitableAttributes.new(model)
+    csl_data = attributes.as_csl_json.stringify_keys
+    citation_item = CiteProc::CitationItem.new(id: csl_data["id"] || "id") do |c|
+      c.data = CiteProc::Item.new(csl_data)
+    end
+    renderer = CiteProc::Ruby::Renderer.new :format => CiteProc::Ruby::Formats::Html.new,
+      :locale => self.class.csl_en_us_locale
+    renderer.render(citation_item, self.class.csl_chicago_style.bibliography).html_safe
+  end
```

Comment:
So this might be "bike shedding", talking about a minor detail that doesn't matter in the end. But humor me for a moment, although it doesn't necessarily need to lead to any code changes. 

If we imagine that we might instantiate a CitationDisplay, and then call `display` on it more than once. (In fact, we have no plans to, which is part of what makes this is a side conversation). 

All this work in this method would be done each time `display` is called -- and it's actually somewhat expensive work. But it will get the exact same result each time, because the result is a function of the work/model passed in, and you can't change the work attribute of a CitationDisplay. So it'll be doing unneeded expensive work more than it needs to. 

The solution is `memoizing`.  Ie, lazily storing in an instance variable. In the most basic way, you could wrap the whole method body in:

```
@display ||= begin
   # existing implementation
end
```. 

Now, the first time it's called, since the @display instance variable is nil, it executes the thing after the `||=`, the begin/end block, and assigns the result to @display. And then returns @display from the `display` method, since it's the last (actually only) line of the method body.  The next time it is called, since @display is not nil, it just returns @display.  (There is a hint about the gotcha here -- if the result of the block _can be nil_, it will never recognize it as already calcualted, and re-calculated it each time. Not sure if that's true here, sometimes it doesn't really matter).   

Now we've lazy cached/memoized the expensive stuff, so it only gets done once no matter how many times the method is called. 

Or, you could go further and extract things like `citeproc_citation_item` and `renderer` to methods, which are themselves memoized. 

```ruby
  def display
    @display ||= renderer.render(citation_item, self.class.csl_chicago_style.bibliography).html_safe
  end

  def citable_attributes
     @citable_attributes ||= CitableAttributes.new(model)
  end

  def csl_citation_item
    @csl_citation_item ||= begin
       csl_data = citable_attributes.as_csl_json.stringify_keys
       citation_item = CiteProc::CitationItem.new(id: csl_data["id"] || "id") do |c|
          c.data = CiteProc::Item.new(csl_data)
       end
    end

    def renderer
       @renderer ||=  CiteProc::Ruby::Renderer.new :format => CiteProc::Ruby::Formats::Html.new,
            :locale => self.class.csl_en_us_locale
     end
```

More readable/maintainable? Less readable/maintainable? Just shuffling deck chairs on the titanic? I think more readable/maintainable (as well as adding memoization), but I could totally be wrong, and this could all just be irrelevant.

Reply:
Yeah, I think memoization is perfectly reasonable. There's a lot of processing here. I'm also getting more used to the idiom. I think this is a great suggestion.

Reply:
Memoized it.

Reply:
Done!